lms/views/predicates/_lti_launch.py

```diff
@@ -123,13 +173,15 @@ def __init__(self, value, config):
         self.canvas_file = CanvasFile(True, config)
         self.url_configured = URLConfigured(True, config)
         self.db_configured = DBConfigured(True, config)
+        self.blackboard_copied = BlackboardCopied(True, config)
 
     def __call__(self, context, request):
         configured = any(
             [
                 self.canvas_file(context, request),
                 self.url_configured(context, request),
                 self.db_configured(context, request),
+                self.blackboard_copied(context, request),
```

Comment:
You can turn on debug logging of SQL queries like this:

```diff
diff --git a/lms/app.py b/lms/app.py
index 613eb4f4..7df719a4 100644
--- a/lms/app.py
+++ b/lms/app.py
@@ -98,4 +98,7 @@ def create_app(global_config, **settings):  # pylint: disable=unused-argument
 
     config.scan()
 
+    import logging
+    logging.getLogger("sqlalchemy.engine").setLevel(logging.DEBUG)
+
     return config.make_wsgi_app()
```

In the case where the `resource_link_id_history` does exist in the DB but the `resource_link_id` doesn't yet (which I think is the worst case) this is what I see (excerpt of the log output):

* The same application instance is `SELECT`'ed twice. (That's a separate issue, but we could cache it in the service layer.)
* The non-existent `ModuleItemConfiguration` for the `resource_link_id` (`_155_1`) is `SELECT`'ed 6 times
* The `ModuleItemConfiguration` for the `resource_link_id_history` (`_144_1`) is `SELECT`'ed 4 times
* A new `ModuleItemConfiguration` for `resource_link_id` is `INSERT`'ed
* There's also a `GroupInfo` `SELECT`, an `LtiLaunches` `INSERT` and a `Course` `SELECT`
* 16 DB queries in total (for the entire LTI launch, not just these views and predicates). This could be reduced to 8 if the `ModuleItemConfiguration` lookups were cached, 7 if the `ApplicationInstance` lookups were also cached

```
BEGIN (implicit)

SELECT application_instances.id ...

SELECT module_item_configurations ...
{'resource_link_id_1': '_155_1', ...

SELECT module_item_configurations ...
{'resource_link_id_1': '_155_1', ...

SELECT module_item_configurations ...
{'resource_link_id_1': '_144_1', ...

SELECT module_item_configurations.id ...
{'resource_link_id_1': '_155_1', ...

SELECT module_item_configurations ...
{'resource_link_id_1': '_155_1', ...

SELECT module_item_configurations ...
{'resource_link_id_1': '_144_1', ...

SELECT module_item_configurations ...
{'resource_link_id_1': '_155_1', ...

SELECT module_item_configurations ...
{'resource_link_id_1': '_144_1', ...

SELECT application_instances.id ...

SELECT module_item_configurations ...
{'resource_link_id_1': '_144_1', ...

SELECT module_item_configurations ...
{'resource_link_id_1': '_155_1', ...

INSERT INTO module_item_configurations ...
{'resource_link_id': '_155_1', ...

SELECT group_info.info ...

INSERT INTO lti_launches ...

SELECT course.consumer_key ...

COMMIT
```

Reply:
Did a quick test of moving `set_document_url()` and `get_document_url()` into a service with `@lru_cache()` and it does indeed reduce the `SELECT module_item_configurations`'s to just two and caching works per request

Reply:
The `ApplicationInstance` `SELECT` is actually [already cached in the services](https://github.com/hypothesis/lms/blob/3dbffc74550e7ba63847c8abb2b5e8152447bc19/lms/services/application_instance_getter.py#L99-L116) but the `LaunchVerifier` service [calls the model directly](https://github.com/hypothesis/lms/blob/3dbffc74550e7ba63847c8abb2b5e8152447bc19/lms/services/launch_verifier.py#L58-L60) instead. That's the reason for the two `ApplicationInstance` lookups. IIRC that was because `LaunchVerifier` couldn't depend on `ApplicationInstanceGetter` because it would create a circular dependency between the services. So if we wanted to optimise that duplicate `ApplicationInstance` `SELECT` away we'd have to resolve that circular dependency in the service layer (I think by splitting out a third service)

Reply:
This has been rebased on top of https://github.com/hypothesis/lms/pull/2269 which adds the caching